node/bft/src/sync/mod.rs

```diff
@@ -515,12 +502,55 @@ impl<N: Network> Sync<N> {
         // conservatively set to the block height minus the maximum number of blocks calculated above.
         // By virtue of the BFT protocol, we can guarantee that all GC range blocks will be loaded.
         let max_gc_height = tip.saturating_sub(max_gc_blocks);
-        let within_gc = (current_height + 1) > max_gc_height;
+        let within_gc = (ledger_height + 1) > max_gc_height;
+
+        if within_gc {
+            // Retrieve the current height, based on the ledger height and the
+            // (unconfirmed) blocks that are already queued up.
+            let start_height = self.compute_sync_height().await;
+
+            // For sanity, update the sync height before starting.
+            // (if this is lower or equal to the current sync height, this is a noop)
+            self.block_sync.set_sync_height(start_height);
+
+            // The height is incremented as blocks are added.
+            let mut current_height = start_height;
+            trace!("Try advancing with block responses (at block {current_height})");
+
+            // If we already were within GC or successfully caught up with GC, try to advance BFT normally again.
+            loop {
+                let next_height = current_height + 1;
+                let Some(block) = self.block_sync.peek_next_block(next_height) else {
+                    break;
+                };
+                info!("Syncing the BFT to block {}...", block.height());
+                // Sync the storage with the block.
+                match self.sync_storage_with_block(block).await {
+                    Ok(_) => {
+                        // Update the current height if sync succeeds.
+                        current_height = next_height;
+                    }
+                    Err(err) => {
+                        // Mark the current height as processed in block_sync.
+                        self.block_sync.remove_block_response(next_height);
```

Comment:
Where do we remove outstanding block responses in the happy Ok path? Do we need to?